cpp/src/io/parquet/experimental/hybrid_scan_impl.hpp

```diff
@@ -0,0 +1,528 @@
+   *
+   * One large contiguous buffer of data allocated and
+   * distributed among the PageInfo structs.
+   */
+  void allocate_level_decode_space();
+
+  /**
+   * @brief Reset the internal state of the reader.
+   */
+  void reset_internal_state();
+
+  /**
+   * @brief Finalize the output table by adding empty columns for the non-selected columns in
+   * schema.
+   *
+   * @tparam RowMaskView View type of the row mask column
+   *
+   * @param[in] read_mode Read mode indicating if we are reading filter or payload columns
+   * @param[in,out] out_metadata The output table metadata
+   * @param[in,out] out_columns The columns for building the output table
+   * @param[in,out] row_mask Boolean column indicating which rows need to be read after page-pruning
+   *                         for filter columns, or after materialize step for payload columns
+   * @return The output table along with columns' metadata
+   */
+  template <typename RowMaskView>
+  cudf::io::table_with_metadata finalize_output(read_mode read_mode,
+                                                table_metadata& out_metadata,
+                                                std::vector<std::unique_ptr<column>>& out_columns,
+                                                RowMaskView row_mask);
+
+  /**
+   * @brief Allocate data buffers for the output columns.
+   *
+   * @param skip_rows Crop all rows below skip_rows
+   * @param num_rows Maximum number of rows to read
+   */
+  void allocate_columns(size_t skip_rows, size_t num_rows);
+
+  /**
+   * @brief Calculate per-page offsets for string data
+   *
+   * @return Vector of total string data sizes for each column
+   */
+  cudf::detail::host_vector<size_t> calculate_page_string_offsets();
+
+  /**
+   * @brief Converts the page data and outputs to columns.
+   *
+   * @param skip_rows Minimum number of rows from start
+   * @param num_rows Number of rows to output
+   */
+  void decode_page_data(size_t skip_rows, size_t num_rows);
+
+  /**
+   * @brief Creates file-wide parquet chunk information.
+   *
+   * Creates information about all chunks in the file, storing it in
+   * the file-wide _file_itm_data structure.
+   */
+  void create_global_chunk_info(cudf::io::parquet_reader_options const& options);
+
+  /**
+   * @brief Computes all of the passes we will perform over the file.
+   */
+  void compute_input_passes();
+
+  /**
+   * @brief Given a set of pages that have had their sizes computed by nesting level and
+   * a limit on total read size, generate a set of {skip_rows, num_rows} pairs representing
+   * a set of reads that will generate output columns of total size <= `chunk_read_limit` bytes.
+   */
+  void compute_output_chunks_for_subpass();
+
+  /**
+   * @brief Check if there is more work to be done.
+   */
+  [[nodiscard]] bool has_more_work() const
+  {
+    return _file_itm_data.num_passes() > 0 &&
+           _file_itm_data._current_input_pass < _file_itm_data.num_passes();
```

Comment:
Can this definition be moved into the .cpp file?

Reply:
Yeah of course. Let me move these